### ISSUE TITLE
v0.5 Bringing back scope_info to Server side

### ISSUE THREAD

**Issue by Mimetis** [Mimetis/Dotmim.Sync#229]
One *new* feature from `v0.4` was to drop the `scope_info` table from the server side.

The main reason was another new feature: In `v0.4` only the client is responsible of the starting point of its own sync session.
Eventually server does not care of any *client start point in time*.
That's why we did not need the `scope_info` table on the server side for handling start and end of sync.
It means a lot of simplicity for handling clients. Because the server is *agnostic* to any clients.

But...

We miss so much information on the server side, that why I've decided to bring the `scope_info` back in the server database.

Even more, the `v0.5` will bring **2** new tables:
- `scope_info_server` : This table will contains the schema declared in your code.
- `scope_info_history`: This table will contains the last sync session from any client, for info.

### scope_info_server

As I said, this table contains some useful informations:
- *sync_scope_name*: Identifier. Each scope name represents a schema
- *sync_scope_schema*: Actual schema

Since we have the schema declared by code, why having schema in database ? 
Well one main reason for that is to be able to recreate portion of the `Dotmim.Sync` sql components when you are making some changes in your coded schema...
Supposing you're adding (or removing) a table from your `SyncSetup`, in your code:

```
// Actually we are using these tables
// var setup = new SyncSetup(new string[] { "Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" });

// Now I need also Product !
var setup = new SyncSetup(new string[] { "Product", "Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" });
```
Today, you have to manually `DeprovisionAsync()` and `ProvisionAsync()` again your whole database.
It could be really complicated...

The idea is to be able to compare your coded schema with the last schema stored in the database.
Then make all mandatories operations in the databse to be sure everything will run smoothly.

- *sync_scope_last_clean_timestamp*: Last time the metadatas for all tables involved in this schema have been deleted.

Now that we can delete metadatas in the tracking tables, we need to be sure that any client who will try to sync its database has its *last time successful* sync time greater than *last cleanup metadatas*

### scope_info_history

This scope table will contains one line per client. Each line will have the last time the client tried to sync.
This table is pretty much the same as the actual `scope_info` table stored on the client:

- *sync_scope_name* : Scope name.
- *scope_last_sync_timestamp*: last **server** timestamp.
- *scope_last_sync_duration*: last **server** duration (Basically how much time to apply client batches and get back server changes)
- *scope_last_sync*: last **server** sync datetime.

FYI, The `scope_info` table from contains the same columns, plus one **client** column representing the client timestamp.

## Your thoughts ?

I'm actually working on this feature. 
But I'm pretty sure I miss some useful information that can be stored in these 2 tables.
What kind of info do you think we should save in these new tables ?

**Comment by lordofffarm** [comment]
Maybe save the migration name or id at the time of provisioning in the scope_info_server?
So that we could compare the current migration and the saved one, and deprov/prov if they are not equal. Obviously, a new migration might not affect the tables that are in the sync schema, so I am not sure if this is the best approach. Maybe as an optional feature?

**Comment by Mimetis** [comment]
Thaks @lordofffarm, that's a good point.
I think having a specific object like a `SyncMigration` or whatever, should be more reliable.
And as you say, the possibility to apply or not the migration.

Thanks for this feedback !

**Comment by mdonogma** [comment]
Suggested features for 0.5 release:
A. Webapi sync options that allow client to flag what tables to sync. Mentioned in previous ticket - this one is critical for anyone serious about offline sync.
B. client or server options to allow multitenant scenario. Mentioned in previous ticket

**Comment by Mimetis** [comment]
This issue has been fixed with the release of the version `v0.5`.
If you still experiencing the same bug, please comment here, I will re-open the issue.
